src/node_messaging.cc

```diff
@@ -544,36 +544,87 @@ std::unique_ptr<MessagePortData> MessagePort::Detach() {
 }
 
 
-void MessagePort::Send(Message&& message) {
-  Mutex::ScopedLock lock(*data_->sibling_mutex_);
-  if (data_->sibling_ == nullptr)
-    return;
-  data_->sibling_->AddToIncomingQueue(std::move(message));
-}
+Maybe<bool> MessagePort::PostMessage(Environment* env,
+                                     Local<Value> message_v,
+                                     Local<Value> transfer_v) {
+  auto isolate = env->isolate();
+  auto obj = object(isolate);
+  auto context = obj->CreationContext();
+
+  std::vector<MessagePort*> transferred_ports;
+
+  // Check if the MessagePort itself is being transferred.
+  if (transfer_v->IsArray()) {
+    Local<Array> transfer_list = transfer_v.As<Array>();
+    uint32_t length = transfer_list->Length();
+    for (uint32_t i = 0; i < length; ++i) {
+      Local<Value> entry;
+      if (!transfer_list->Get(context, i).ToLocal(&entry))
+        return Nothing<bool>();
+
+      if (env->message_port_constructor_template()->HasInstance(entry)) {
+        if (entry == obj) {
+          Local<Value> argv[] = {
+            FIXED_ONE_BYTE_STRING(isolate,
+                                  "Transfer list contains source port"),
+            FIXED_ONE_BYTE_STRING(isolate, "DataCloneError")
+          };
+          env->isolate()->ThrowException(
+              env->domexception_function()
+                  ->NewInstance(context, arraysize(argv), argv)
+                  .ToLocalChecked());
+          return Nothing<bool>();
+        }
+        MessagePort* port = Unwrap<MessagePort>(entry.As<Object>());
+        CHECK_NE(port, nullptr);
+        transferred_ports.push_back(port);
+      }
+    }
+  }
```

Comment:
Can we integrate this with the loop in `Message::Serialize`? There’s a lot of common code, and I’m not sure we really need to unpack the list twice?

Reply:
I actually had that approach initially, but then I realized that I would need to access `MessagePort` and/or `MessagePortData` internals from `Message`. I could of course add a `friend` but I decided that object serialization should be independent of the channel communication. Also, aligning with the HTML Standard (whose algorithm is written in this style) would be good.

Reply:
I realize that it’s tricky, but I’d really like to merge these … if you want, I can try to figure out something myself?

I think it we can make it work if we added a `MessagePort* source` argument to `Serialize()` that we check against, and add a getter for `Message::message_ports_` that we check after serialization?

Reply:
Also, at least Firefox and Chrome invoke getters for `transferList` only once, as an extra data point

Reply:
@addaleax 

> I think it we can make it work if we added a `MessagePort* source` argument to `Serialize()` that we check against, and add a getter for `Message::message_ports_` that we check after serialization?

Sure, that would work. However, checking things after serialization would mean `ArrayBuffer`s have already been neutered, etc.

> Also, at least Firefox and Chrome invoke getters for transferList only once, as an extra data point

Yeah, browsers usually have an IDL conversion layer, and use native data types for all internal operations. We could do the same and convert `transferList` to a `std::vector`.

Reply:
> > I think it we can make it work if we added a `MessagePort* source` argument to `Serialize()` that we check against, and add a getter for `Message::message_ports_` that we check after serialization?

> Sure, that would work. However, checking things after serialization would mean `ArrayBuffer`s have already been neutered, etc.

To clarify, the first check (for `source`) could still happen before the actual serialization, while still unpacking the transfer list. The second check would happen afterwards, but that’s also what this PR currently does (and intentionally, if I read the added `test-worker-message-port-transfer-target.js` correctly?)